src/client/Client.cc

```diff
@@ -7550,6 +7550,21 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (in->caps_issued_mask(CEPH_CAP_FILE_EXCL)) {
+    if ((mask & CEPH_SETATTR_SIZE) && !(mask & CEPH_SETATTR_KILL_SGUID)) {
```

Comment:
FWIW, I suspect the handling of CEPH_SETATTR_KILL_SGUID is probably racy as _write() doesn't ensure we have As caps before checking whether the mode has setuid or setgid set. That's not really a problem with your patch though, just an observation. I think I did it that way because it was going to suck for performance otherwise. Oh well.

Reply:
The `_write()` will make sure that the `As` caps has been issued before checking these two bits in `in->mode`.